Port existing heuristics to TD framework

ZainRizvi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107071


This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

ZainRizvi committed:
--no-edit

[ghstack-poisoned]

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

huydhn commented on tools/testing/target_determination/heuristics/utils.py:
```diff
@@ -0,0 +1,41 @@
+from pathlib import Path
+from typing import List, Set
+
+REPO_ROOT = Path(__file__).resolve().parent.parent.parent.parent.parent
```
This line is so lol

ZainRizvi replied:
🫣 (this could really use a more generic helper function...)

clee2000 commented on tools/testing/target_determination/heuristics/edited_by_pr.py:
```diff
@@ -0,0 +1,34 @@
+    def get_test_priorities(self, tests: List[str]) -> TestPrioritizations:
+        critical_tests = _get_modified_tests()
+        test_rankings = TestPrioritizations()
+        test_rankings.highly_relevant = list(critical_tests)
```
add a sorted here

clee2000 commented on tools/testing/target_determination/heuristics/previously_failed_in_pr.py:
```diff
@@ -0,0 +1,54 @@
+    def get_test_priorities(self, tests: List[str]) -> TestPrioritizations:
+        critical_tests = _get_previously_failing_tests()
+        test_rankings = TestPrioritizations()
+        test_rankings.highly_relevant = list(critical_tests)
```
add a sorted here

huydhn commented on tools/testing/target_determination/heuristics/__init__.py:
```diff
@@ -1,21 +1,22 @@
-    # CorrelatedWithHistoricalFailures(),
+    PreviouslyFailedInPR(),
+    EditedByPR(),
+    CorrelatedWithHistoricalFailures(),
```
As this is a big PR to roll out, what do you think about holding `CorrelatedWithHistoricalFailures` heuristic for now and just go ahead with `PreviouslyFailedInPR` and `EditedByPR`, which are existing features?  Monitoring it for a few days, then `CorrelatedWithHistoricalFailures` will follow

ZainRizvi replied:
`CorrelatedWithHistoricalFailures` is also an existing feature, I just gave it a new name. You'll see it called `_get_file_rating_tests` in `test_selections.py`

huydhn replied:
Ah ok, I thought that Cat has reverted and still working on that.  Sounds good if that changes has been landed again.

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

clee2000 commented on tools/testing/target_determination/heuristics/edited_by_pr.py:
```diff
@@ -0,0 +1,35 @@
+
+    # Tests must always be returned in a deterministic order.
+    # Otherwise it breaks our test sharding logic
+    return python_test_file_to_test_name(set(changed_files))
```
Does set have an ordering in python?

ZainRizvi replied:
no, but this comment was in the wrong place. I sort the return value from this function

clee2000 commented on tools/testing/target_determination/heuristics/interface.py:
```diff
@@ -41,24 +56,84 @@ def integrate_priorities(self, other: "TestPrioritizations") -> None:
+            for test in unexpected:
+                print(f"  {test}")
+        else:
+            print("No unexpected tests found")
```
In what cases would you expect conflicts or unexpected tests?

ZainRizvi replied:
my apologies, I thought I'd deleted that code (it was for debugging). Turns out the files were sitting uncommitted on my box

clee2000 commented on tools/testing/target_determination/heuristics/previously_failed_in_pr.py:
```diff
@@ -0,0 +1,56 @@
+
+    prioritized_tests = _parse_prev_failing_test_files(last_failed_tests)
+
+    return python_test_file_to_test_name(prioritized_tests)
```
Same question as above with set ordering

ZainRizvi replied:
same ans

clee2000 reviewed (commented):
I think you might have to rebase again sorry

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

ZainRizvi committed:
Update on "Port existing heuristics to TD framework"



This PR looks big, but it's mostly just refactorings with a bit of dead code deletion. Exceptions are:
- Some metric emissions were changed to comply with the new TD format
- Some logging changes 
- We now run tests in three batches (highly_relevant, probably_relevant, unranked_relevance) instead of the previous two (prioritized and general)
  
Refactorings done:
- Moves all test reordering code to the new TD framework
- Refactors run_test.py to cleanly support multiple levels of test priorities
- Deletes some dead code that was originally written for logging

[ghstack-poisoned]

ZainRizvi commented:
@pytorchbot merge
